### PR TITLE
Make `Replay.Frames` immutable

### PR DESCRIPTION
**Pull request by ekrctb**
The frames of a replay must be sorted by frame time to make sense. However, it was easy to break the invariant as the `Replay` class exposed a `List<ReplayFrame>` directly.
In this PR, I change the type of `Replay.Frames` to `IReadonlyList<ReplayFrame>` to prevent accidental breaking of the invariant.

The `Replay` class has a constructor to take an enumerable of frames, which will be automatically sorted. In this way, the frame order invariant is automatically satisfied, and most existing uses of the `Replay` class in rulesets can be adopted with minimal changes.

The `StreamingReplay` class is introduced. This class supports an extension of the replay by adding more frames at the end. Once it is marked as complete, no more frames can be added. It is used by the spectator and the replay recording.
The `StreamingReplay : Replay` inheritance is a bit strange in terms of the substitution principle. If the list of frames of a `Replay` is expected to be immutable then the expectation is broken when a `StreamingReplay` is given. Therefore `Replay` type can only promise invariant that is also true in `StreamingReplay`.
I choose this `StreamingReplay : Replay` way to minimize breaking changes of the ruleset API.

### CODE REVIEW

**Review by frenzibyte — COMMENTED** [review]
Cannot test spectator for some reason, but this seems a sensible direction for me either way. Just have a few comments.

File: osu.Game/Rulesets/UI/ReplayRecorder.cs [hunk 1]
```diff
@@ -36,7 +38,11 @@ public abstract class ReplayRecorder<T> : ReplayRecorder, IKeyBindingHandler<T>
 
         protected ReplayRecorder(Score target)
         {
+            if (!(target.Replay is StreamingReplay streamingReplay))
+                throw new InvalidOperationException($"{nameof(ReplayRecorder)} cannot target a non-{nameof(StreamingReplay)} replay.");
+
```
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Rulesets/UI/ReplayRecorder.cs:43]: That would just be papering over the big flaw in this change which was already pointed out in the OP by @ekrctb. Namely that the streaming replay violates the substitution principle and therefore leads to all types of weird checks.

To me that's the single biggest problem with this PR. `StreamingReplay : Replay` should be at most temporary/transitional IMO. They need to be separate types, and all inheritors of `Replay` need to be immutable. If we've gone down the road of having the replay be read-only then I'd rather see more breaking changes than what's currently being done.

File: osu.Game.Rulesets.Osu/Replays/OsuAutoGeneratorBase.cs [hunk 2]
```diff
@@ -28,15 +27,12 @@ public abstract class OsuAutoGeneratorBase : AutoGenerator
...
 
-        protected Replay Replay;
-        protected List<ReplayFrame> Frames => Replay.Frames;
+        protected readonly List<ReplayFrame> Frames = new List<ReplayFrame>();
         private readonly IReadOnlyList<IApplicableToRate> timeAffectingMods;
 
         protected OsuAutoGeneratorBase(IBeatmap beatmap, IReadOnlyList<Mod> mods)
             : base(beatmap)
         {
-            Replay = new Replay();
```
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game.Rulesets.Osu/Replays/OsuAutoGeneratorBase.cs:38]: The refactoring you suggest has been purposefully omitted from #12701 and is out of scope here.

**Comment by bdach** [comment]
As stated above, I can't see myself agreeing with this change with the way `StreamingReplay` is set up currently. It makes things actively more confusing and less type-safe. I don't buy the argument that it's a less breaking change. It still breaks things (see a few runtime exceptions added here), and arguably in a worse, more opaque way.

File: osu.Game.Rulesets.Osu/Replays/OsuAutoGeneratorBase.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game.Rulesets.Osu/Replays/OsuAutoGeneratorBase.cs:38]: I see, missed noticing that, would be out of scope in that case so fair enough.

**Comment by smoogipoo** [comment]
As we discussed before, maybe the builder pattern works better here? Something like the following?:
```csharp
public interface IReplay
{
    IReadOnlyList<ReplayFrame> Frames { get; }
    bool IsComplete { get; }
}

public class Replay : IReplay
{
    public IReadOnlyList<ReplayFrame> Frames { get; }
    public bool IsComplete => true;
}

public class ReplayBuilder
{
    private readonly MutableReplay replay = new MutableReplay();

    public void AddFrame(ReplayFrame frame) => replay.Frames.Add(frame);
    public void SetComplete() => replay.IsComplete = true;

    public IReplay ToReplay() => replay;

    private class MutableReplay : IReplay
    {
        public readonly List<ReplayFrame> Frames = new List<ReplayFrame>();
        public bool IsComplete;

        IReadOnlyList<ReplayFrame> IReplay.Frames => Frames;
        bool IReplay.IsComplete => IsComplete;
    }
}

public class ReplayFrame
{
}
```

**Comment by ekrctb** [comment]
One question is that the interface `IReplay` has very ad-hoc requirement that is basically "what `FramedReplayInputHandler` wants". I think the generic `IReplay` name is not perfectly apt for this purpose.
Probably `FramedReplayInputHandler` API should be changed to something like:

```csharp
void Init(<Immutable>Replay);
void AddFrame(ReplayFrame frame);
void SetComplete();
```

Then `StreamingReplay` is not needed anymore.

**Comment by bdach** [comment]
I definitely agree with the paths above more, although they still seem a bit overdone. I'm not sure why `SetComplete`/`IsComplete` are popping up in the proposals above still. What am I missing? I'd honestly just do

```csharp
public class Replay
{
	public IReadOnlyList<ReplayFrame> Frames { get; }

	public Replay(IEnumerable<ReplayFrame> frames)
	{
		Frames = frames.OrderBy(f => f.Time).ToList();
	}
}

public class ReplayBuilder
{
	public void AddFrame(ReplayFrame frame);
	public Replay Build(); // is just ctor call at this point, so may not even be worth the hassle
}
```

Note that the builder may not even be strictly needed at that point anymore, but may be a decent convenience thing for backwards compatibility.

**Comment by ribbanya** [comment]
I feel `StreamingReplay` unnecessarily deviates from the well-established [writer](https://docs.microsoft.com/en-us/dotnet/api/system.io.streamwriter?view=net-5.0) pattern which has `Close`, `Write`, and an `IDisposable` implementation.

I see no reason it should inherit from `Replay` and hijack its supposedly immutable frames. Something like this:

```cs
public class ReplayWriter : IDisposable /* optional collection interfaces to access frames */ {

    // Not necessarily the most performant implementation, this is just a rough example
    private readonly SortedList<double, ReplayFrame> frames;
    private bool disposed;

    public ReplayWriter(int capacity) {
        this.frames = new SortedList<double, ReplayFrame>(capacity);
    }

    public ReplayWriter(IEnumerable<ReplayFrame> frames) {
        this.frames = new SortedList<double, ReplayFrame>();
        Write(frames);
    }

    void IDisposable.Dispose() {
        this.disposed = true;
    }

    public void Close() {
        if (CheckDisposed(out var ex)) throw ex;

        Dispose();
    }

    private bool CheckDisposed(out Exception ex) {
        if (this.disposed) {
            ex = new InvalidOperationException("Writer is closed.");
            return true;
        }
        ex = null;
        return false;
    }

    private void Add(ReplayFrame frame) => this.frames.Add(frame.Time, frame);

    public void Write(ReplayFrame frame) {
        if (CheckDisposed(out var ex)) throw ex;

        Add(frame);
    }

    public void Write(IEnumerable<ReplayFrame> frames) {
        if (CheckDisposed(out var ex)) throw ex;

        foreach (var frame in frames) {
            Add(frame);
        }
    }

    public void ToReplay() => new Replay(this.frames);

    public void CopyTo(List<ReplayFrame> buffer) {
        buffer.Clear();
        buffer.AddRange(this.frames);
    }
}
```

**Comment by bdach** [comment]
> I feel `StreamingReplay` unnecessarily deviates from the well-established writer pattern which has `Close`, `Write`, and an `IDisposable` implementation.

That's because it doesn't fit the pattern. There's no actual "streaming" there, the proposal is just badly named. There's no need for `IDisposable` either, it's just a memory buffer.

> To be honest, giving it more thought, I don't really see the need for a class on top of `SortedList` at all.

[I proposed this already (see first point)](https://github.com/ppy/osu/pull/12701#issuecomment-833579165). But things don't seem to be going down that way.

> `ReplayFrame` is mutable, including its `Time` (a public field).

Fair point, although probably one of the lowest on the list of issues to address.

**Comment by ribbanya** [comment]
> Fair point, although probably one of the lowest on the list of issues to address.

Yeah, my point was more that you can do a lot of things to try to prevent programmers from writing bad code, but that's what tests, reviews, and documentation are for and not things like `IReadOnlyList<T>`.

It makes perfect sense to me that `ReplayFrame` is a mutable reference type. I think the real solution is to just add some sanity checking methods to `Replay` that tells it to sort and/or prune frames a maximum of once per update loop, or something.

You could use C# events, dirty flags, or indeed just a sorted data structure as you mention (though it would have to understand underlying changes in frames).

**Comment by ekrctb** [comment]
Close as not enough interests and this PR doesn't solve real issues.
To restate to be referenced in future, the issue (even if this PR is applied) is what `FramedReplayInputHandler` is expecting to the `Replay` type is very ad-hoc:

- Frames must be sorted by time.
- If the frames are inserted/deleted arbitrary, the invariant is broken.
- But it can handle appending frames at the end, as long as `HasReceivedAllFrames` is set correctly (can break if `HasReceivedAllFrames` is set incorrectly).

I imagine `FramedReplayInputHandler` should not store a `Replay`, but only an internal list of frames. Probably also `AddFrame` for the replay streaming purpose.

**Pull request closed without merging**